### PR TITLE
Add merchant authentication using Firebase

### PR DESCRIPTION
**Pull request by karenfrilya97**

### CODE REVIEW

File: server/tsconfig.json [hunk 1]
```diff
@@ -12,7 +12,10 @@
     "outDir": "build",
     "skipLibCheck": true,
     "strict": true,
-    "baseUrl": "./src/"
+    "baseUrl": "./src/",
+    "typeRoots": [
+      "./custom_typings"
```
Comment (hzxcaryn) [inline server/tsconfig.json:17]: Should we be consistent and use `-` case for file and folder names? :p

File: server/src/middleware/merchant-auth.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Request, Response, NextFunction} from 'express';
+import admin from 'firebase-admin';
+
+admin.initializeApp();
```
Comment (hzxcaryn) [inline server/src/middleware/merchant-auth.ts:20]: Should we do something similar to the client side, creating a `src/firebase.ts` which initializes the app and exports the `const auth = admin.auth();`?

File: server/custom_typings/express/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare namespace Express {
+  interface Request {
+    decoded: import('firebase-admin').auth.DecodedIdToken;
```
Comment (ifhuang) [inline server/custom_typings/express/index.d.ts:19]: decoded -> firebaseDecodedIdToken?

File: server/src/middleware/merchant-auth.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+const merchantAuth = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    if (!req.headers.authorization) throw new Error('Unauthorized');
+
+    const [_, idToken] = req.headers.authorization.split(' ');
```
Comment (ifhuang) [inline server/src/middleware/merchant-auth.ts:30]: idToken -> firebaseIdToken?

File: server/tsconfig.json (hunk 1 above)
Comment (karenfrilya97) [reply server/tsconfig.json:17]: I was following the node_modules cos this one is the same level as that. But I guess we can just standardise all the files that we make.

File: server/src/middleware/merchant-auth.ts (hunk 2 above)
Comment (karenfrilya97) [reply server/src/middleware/merchant-auth.ts:20]: Yep, I think I've already done that  https://github.com/googleinterns/gpay-group-buy/pull/54/commits/2f1b38df7a3993c6eb523b4b94badb447d99ecd2

File: server/tsconfig.json (hunk 1 above)
Comment (karenfrilya97) [reply server/tsconfig.json]: Also, I realise I can put this folder inside src then it'll be better.

File: server/custom_typings/express/index.d.ts (hunk 3 above)
Comment (karenfrilya97) [reply server/custom_typings/express/index.d.ts]: I think the convention is `req.decoded` when you put something you decoded from a token.

File: server/src/middleware/merchant-auth.ts (hunk 4 above)
Comment (karenfrilya97) [reply server/src/middleware/merchant-auth.ts:30]: Okay!

File: server/src/middleware/merchant-auth.ts (hunk 2 above)
Comment (hzxcaryn) [reply server/src/middleware/merchant-auth.ts:20]: @karenfrilya97 Isn't that on the client side? I actually meant like doing the same for server side too

**Pull request merged**